src/spi/clone/CachingDirectoryManager.ts

```diff
@@ -60,13 +67,116 @@ export const CachingDirectoryManager: DirectoryManager = {
                 }
             }));
     },
+
+    reportMetrics(): CachingOfClonesMetrics {
+        return CachingOfClonesMetricsCacheImpl.report();
+    },
 };
 
+/*
+ * Metrics reporting
+ */
+
+const sleepPlease: (timeout: number) => Promise<void> =
+    promisify((a, b) => setTimeout(b, a));
+
+const MetricLogFrequency = 1000 * 60 * 10; // ten minutes
+
+/* This returns a promise that never finishes, unless it throws somehow
+ */
+export function logCachingDirectoryManagerMetricsAllDay(): Promise<void> {
+    return sleepPlease(MetricLogFrequency)
+        .then(() => logger.info(CachingOfClonesMetricsCacheImpl.report().print()))
+        .then(() => logCachingDirectoryManagerMetricsAllDay());
+}
+
+export interface PerRepoCachingMetrics {
+    reuses: number;
+    fallbacks: number;
+}
+
+class CachingOfClonesMetricsCache {
+
+    private reuseCounts: any = {};
+    private fallbackCounts: any = {};
+
+    public report() {
+        // this makes a copy each time so it doesn't update underneath you
+        // mainly useful in testing
+        return new CachingOfClonesMetrics(this.reuseCounts, this.fallbackCounts);
+    }
+
+    public incrementReuses(repoId): void {
+        if (!this.reuseCounts[keyFor(repoId)]) {
+            this.reuseCounts[keyFor(repoId)] = 1;
+            return;
+        }
+        this.reuseCounts[keyFor(repoId)]++;
+    }
+
+    public incrementFallbacks(repoId): void {
+        if (!this.fallbackCounts[keyFor(repoId)]) {
+            this.fallbackCounts[keyFor(repoId)] = 1;
+            return;
+        }
+        this.fallbackCounts[keyFor(repoId)]++;
+    }
+
+}
+
+function keyFor(repoId: RepoId) {
+    return `${repoId.owner}/${repoId.repo}`;
```

Comment:
What should we do about owner/repo name collisions across GHE instances?

Reply:
I agree that's a thing and we're not handling it -- we're not handling GHE in tons of the client code.